File: libs/python/mail_runtime_core/package_tracking.py
#!/usr/bin/env python3
"""Shared mail-to-package-tracking helpers built on package_tracking_core."""

from __future__ import annotations

import re
from typing import Callable, Protocol

import package_tracking_core

from .runtime import MailEnvelope


DELIVERY_KEYWORDS = [
    "delivered",
    "package delivered",
    "your order has been delivered",
    "delivery complete",
    "successfully delivered",
    "has been delivered",
    "your package was delivered",
    "item delivered",
    "order delivered",
]

_AMAZON_DOMAINS = ("amazon.com", "amazonlogistics.com")
_NARVAR_URL_PATTERN = r"https?://[^\s\"'<>]*narvar\.com/[^\s\"'<>]*"


class TrackingClient(Protocol):
    """Minimal tracking client contract used by mail actions."""

    def is_shipping_sender(self, sender_email: str) -> bool: ...

    def scan_text_for_tracking_numbers(self, text: str) -> list[dict]: ...

    def extract_tracking_from_urls(self, text: str) -> list[dict]: ...

    def fetch_narvar_tracking(self, url: str) -> list[dict]: ...

    def add_package(self, tracking_number: str, carrier: str, label: str) -> dict: ...

    def remove_package(self, tracking_number: str) -> dict: ...


def is_delivery_notification(subject: str) -> bool:
    """Return True when the subject suggests a package was delivered."""

    low = (subject or "").lower()
    return any(keyword in low for keyword in DELIVERY_KEYWORDS)


def load_tracking_client() -> TrackingClient:
    """Load and return the shared package tracking core module."""

    return package_tracking_core


def _combined_body(envelope: MailEnvelope) -> str:
    return " ".join(filter(None, [envelope.body_text, envelope.body_html]))


def _is_amazon_sender(sender_email: str) -> bool:
    sender_email_low = (sender_email or "").lower()
    return any(
        sender_email_low.endswith("@" + domain)
        or bool(
            re.search(
                r"@(?:[a-z0-9-]+\.)*" + re.escape(domain) + r"$",
                sender_email_low,
            )
        )
        for domain in _AMAZON_DOMAINS
    )


def scan_and_add_packages(
    envelope: MailEnvelope,
    *,
    account_label: str,
    logger: Callable[[str], None],
    tracking_client_loader: Callable[[], TrackingClient] = load_tracking_client,
) -> list[str]:
    """Scan an email envelope for tracking numbers and add them to package tracking."""

    sender_email = envelope.sender_email or "unknown"
    sender_name = envelope.sender_name or ""
    subject = envelope.subject or "(no subject)"

    try:
        tracking_client = tracking_client_loader()

        if not tracking_client.is_shipping_sender(sender_email):
            logger(f"skipping tracking scan: non-shipping sender {sender_email}")
            return []

        if _is_amazon_sender(sender_email):
            logger(
                f"skipping tracking scan: Amazon sender {sender_email} "
                "(not trackable externally)"
            )
            return []

        body_text = envelope.body_text or ""
        found = tracking_client.scan_text_for_tracking_numbers(body_text) if body_text else []

        combined_text = _combined_body(envelope)
        url_found = tracking_client.extract_tracking_from_urls(combined_text)

        narvar_urls = re.findall(_NARVAR_URL_PATTERN, combined_text, re.IGNORECASE)
        for narvar_url in narvar_urls[:3]:
            url_found.extend(tracking_client.fetch_narvar_tracking(narvar_url))

        seen_numbers = {result["tracking_number"] for result in found}
        for result in url_found:
            tracking_number = result["tracking_number"]
            if tracking_number not in seen_numbers:
                seen_numbers.add(tracking_number)
                found.append(result)

        if not found:
            return []

        added: list[str] = []
        for tracking_info in found:
            tracking_number = tracking_info["tracking_number"]
            carrier = tracking_info["carrier"]
            label = f"{account_label}: {sender_name or sender_email} - {subject[:40]}"

            result = tracking_client.add_package(tracking_number, carrier, label)
            if "error" in result:
                logger(
                    f"warn: failed to add package {tracking_number}: {result.get('error')}"
                )
                continue

            added.append(tracking_number)
            logger(f"📦 added package: {tracking_number} ({carrier}) — {label}")

        return added
    except Exception as exc:
        logger(f"error: package tracking failed: {exc}")
        return []


def scan_and_remove_delivered(
    envelope: MailEnvelope,
    *,
    logger: Callable[[str], None],
    tracking_client_loader: Callable[[], TrackingClient] = load_tracking_client,
) -> list[str]:
    """Scan a delivery email and remove matching tracked packages."""

    scan_text = envelope.body_text or envelope.subject or ""

    try:
        tracking_client = tracking_client_loader()
        found = tracking_client.scan_text_for_tracking_numbers(scan_text)
        if not found:
            logger(f"delivery email but no tracking number found: {envelope.subject}")
            return []

        removed: list[str] = []
        for tracking_info in found:
            tracking_number = tracking_info["tracking_number"]
            carrier = tracking_info["carrier"]
            result = tracking_client.remove_package(tracking_number)
            if result.get("success"):
                removed.append(tracking_number)
                logger(f"✅ removed delivered package: {tracking_number} ({carrier})")
            elif result.get("error") == "not_found":
                logger(
                    f"delivery notice for untracked package: {tracking_number} — ignoring"
                )
            else:
                logger(f"warn: failed to remove {tracking_number}: {result}")

        return removed
    except Exception as exc:
        logger(f"error: delivery removal failed: {exc}")
        return []

File: libs/ts/mail_runtime_core/src/package-tracking.ts
/**
 * Shared mail-to-package-tracking helpers built on package_tracking_core.
 * TS port of mail_runtime_core/package_tracking.py
 */

import * as packageTrackingCore from "@openclaw/package-tracking-core";
import type { MailEnvelope } from "./runtime.js";

// ---------------------------------------------------------------------------
// TrackingClient interface
// ---------------------------------------------------------------------------

export interface TrackingClient {
  isShippingSender(senderEmail: string): boolean;
  scanTextForTrackingNumbers(text: string): Array<{ tracking_number: string; carrier: string }>;
  extractTrackingFromUrls(text: string): Array<{ tracking_number: string; carrier: string }>;
  fetchNarvarTracking(url: string): Array<{ tracking_number: string; carrier: string }>;
  addPackage(trackingNumber: string, carrier: string, label: string): Record<string, unknown>;
  removePackage(trackingNumber: string): Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const DELIVERY_KEYWORDS = [
  "delivered",
  "package delivered",
  "your order has been delivered",
  "delivery complete",
  "successfully delivered",
  "has been delivered",
  "your package was delivered",
  "item delivered",
  "order delivered",
];

const AMAZON_DOMAINS = ["amazon.com", "amazonlogistics.com"];
const NARVAR_URL_PATTERN = /https?:\/\/[^\s"'<>]*narvar\.com\/[^\s"'<>]*/gi;

// ---------------------------------------------------------------------------
// isDeliveryNotification
// ---------------------------------------------------------------------------

export function isDeliveryNotification(subject: string | null | undefined): boolean {
  const low = (subject ?? "").toLowerCase();
  return DELIVERY_KEYWORDS.some((kw) => low.includes(kw));
}

// ---------------------------------------------------------------------------
// loadTrackingClient
// ---------------------------------------------------------------------------

export function loadTrackingClient(): TrackingClient {
  return packageTrackingCore as unknown as TrackingClient;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function combinedBody(envelope: MailEnvelope): string {
  return [envelope.body_text, envelope.body_html].filter(Boolean).join(" ");
}

function isAmazonSender(senderEmail: string): boolean {
  const low = (senderEmail ?? "").toLowerCase();
  return AMAZON_DOMAINS.some(
    (domain) =>
      low.endsWith("@" + domain) ||
      new RegExp(`@(?:[a-z0-9-]+\\.)*${domain.replace(/\./g, "\\.")}$`).test(low),
  );
}

// ---------------------------------------------------------------------------
// scanAndAddPackages
// ---------------------------------------------------------------------------

export async function scanAndAddPackages(
  envelope: MailEnvelope,
  options: {
    accountLabel: string;
    logger: (msg: string) => void;
    trackingClientLoader?: () => TrackingClient;
  },
): Promise<string[]> {
  const { accountLabel, logger, trackingClientLoader = loadTrackingClient } = options;
  const senderEmail = envelope.sender_email || "unknown";
  const senderName = envelope.sender_name || "";
  const subject = envelope.subject || "(no subject)";

  try {
    const trackingClient = trackingClientLoader();

    if (!trackingClient.isShippingSender(senderEmail)) {
      logger(`skipping tracking scan: non-shipping sender ${senderEmail}`);
      return [];
    }

    if (isAmazonSender(senderEmail)) {
      logger(
        `skipping tracking scan: Amazon sender ${senderEmail} (not trackable externally)`,
      );
      return [];
    }

    const bodyText = envelope.body_text || "";
    const found = bodyText
      ? trackingClient.scanTextForTrackingNumbers(bodyText)
      : [];

    const combined = combinedBody(envelope);
    const urlFound = trackingClient.extractTrackingFromUrls(combined);

    const narvarUrls = combined.match(NARVAR_URL_PATTERN) ?? [];
    for (const narvarUrl of narvarUrls.slice(0, 3)) {
      urlFound.push(...trackingClient.fetchNarvarTracking(narvarUrl));
    }

    const seenNumbers = new Set(found.map((r) => r.tracking_number));
    for (const result of urlFound) {
      if (!seenNumbers.has(result.tracking_number)) {
        seenNumbers.add(result.tracking_number);
        found.push(result);
      }
    }

    if (found.length === 0) return [];

    const added: string[] = [];
    for (const trackingInfo of found) {
      const { tracking_number: trackingNumber, carrier } = trackingInfo;
      const label = `${accountLabel}: ${senderName || senderEmail} - ${subject.slice(0, 40)}`;
      const result = trackingClient.addPackage(trackingNumber, carrier, label);
      if ("error" in result) {
        logger(`warn: failed to add package ${trackingNumber}: ${result["error"]}`);
        continue;
      }
      added.push(trackingNumber);
      logger(`📦 added package: ${trackingNumber} (${carrier}) — ${label}`);
    }

    return added;
  } catch (exc) {
    logger(`error: package tracking failed: ${exc}`);
    return [];
  }
}

// ---------------------------------------------------------------------------
// scanAndRemoveDelivered
// ---------------------------------------------------------------------------

export async function scanAndRemoveDelivered(
  envelope: MailEnvelope,
  options: {
    logger: (msg: string) => void;
    trackingClientLoader?: () => TrackingClient;
  },
): Promise<string[]> {
  const { logger, trackingClientLoader = loadTrackingClient } = options;
  const scanText = envelope.body_text || envelope.subject || "";

  try {
    const trackingClient = trackingClientLoader();
    const found = trackingClient.scanTextForTrackingNumbers(scanText);
    if (found.length === 0) {
      logger(`delivery email but no tracking number found: ${envelope.subject}`);
      return [];
    }

    const removed: string[] = [];
    for (const trackingInfo of found) {
      const { tracking_number: trackingNumber, carrier } = trackingInfo;
      const result = trackingClient.removePackage(trackingNumber);
      if ((result as Record<string, unknown>)["success"]) {
        removed.push(trackingNumber);
        logger(`✅ removed delivered package: ${trackingNumber} (${carrier})`);
      } else if ((result as Record<string, unknown>)["error"] === "not_found") {
        logger(`delivery notice for untracked package: ${trackingNumber} — ignoring`);
      } else {
        logger(`warn: failed to remove ${trackingNumber}: ${JSON.stringify(result)}`);
      }
    }

    return removed;
  } catch (exc) {
    logger(`error: delivery removal failed: ${exc}`);
    return [];
  }
}
